src/main/webapp/app/exercises/shared/feedback/feedback.service.ts

```diff
@@ -0,0 +1,188 @@
+import { StaticCodeAnalysisIssue } from 'app/entities/static-code-analysis-issue.model';
+
+@Injectable({ providedIn: 'root' })
+export class FeedbackService {
+    constructor(private translateService: TranslateService) {}
+
+    /**
+     * Creates a feedback item with a category, title and text for each feedback object.
+     * @param feedbacks The list of feedback objects.
+     * @param isProgrammingExercise
+     * @param showTestDetails
+     */
+    public createFeedbackItems(feedbacks: Feedback[], isProgrammingExercise: boolean, showTestDetails: boolean): FeedbackItem[] {
+        if (isProgrammingExercise) {
+            return feedbacks.map((feedback) => this.createProgrammingExerciseFeedbackItem(feedback, showTestDetails));
+        } else {
+            return feedbacks.map((feedback) => ({
+                type: FeedbackItemType.Feedback,
+                category: this.translateService.instant('artemisApp.result.detail.feedback'),
+                title: feedback.text,
+                text: feedback.detailText,
+                previewText: computeFeedbackPreviewText(feedback.detailText),
+                positive: feedback.positive,
+                credits: feedback.credits,
+            }));
+        }
+    }
+
+    /**
+     * Creates a feedback item for feedback received for a programming exercise.
+     * @param feedback The feedback from which the feedback item should be created.
+     * @param showTestDetails
+     * @private
+     */
+    private createProgrammingExerciseFeedbackItem(feedback: Feedback, showTestDetails: boolean): FeedbackItem {
+        const previewText = computeFeedbackPreviewText(feedback.detailText);
+
+        if (Feedback.isSubmissionPolicyFeedback(feedback)) {
+            return this.createProgrammingExerciseSubmissionPolicyFeedbackItem(feedback, previewText);
+        } else if (Feedback.isStaticCodeAnalysisFeedback(feedback)) {
+            return this.createProgrammingExerciseScaFeedbackItem(feedback, showTestDetails);
+        } else if (feedback.type === FeedbackType.AUTOMATIC) {
+            return this.createProgrammingExerciseAutomaticFeedbackItem(feedback, showTestDetails, previewText);
+        } else if ((feedback.type === FeedbackType.MANUAL || feedback.type === FeedbackType.MANUAL_UNREFERENCED) && feedback.gradingInstruction) {
+            return this.createProgrammingExerciseGradingInstructionFeedbackItem(feedback, showTestDetails, previewText);
+        } else {
+            return this.createProgrammingExerciseTutorFeedbackItem(feedback, showTestDetails, previewText);
+        }
+    }
+
+    /**
+     * Creates a feedback item from a submission policy feedback.
+     * @param feedback The submission policy feedback.
+     * @param previewText The preview text for the feedback item.
+     * @private
+     */
+    private createProgrammingExerciseSubmissionPolicyFeedbackItem(feedback: Feedback, previewText?: string): FeedbackItem {
```

Comment:
Any suggestions how to refactor `createProgrammingExercise${SPECIFIC_INSTANCE}()` methods?

Reply:
Thanks for the suggestion! I decided to change the `FeedbackItemType` in general as its current implementation is not useful right now. In doing so I will try to come up with some inheritance based structure, rendering a refactoring now unnecessary.